src/pybind/mgr/diskprediction/agent/predict/prediction.py

```diff
@@ -0,0 +1,308 @@
+                disk_type = 7
+            else:
+                disk_type = 1
+        return disk_type
+
+    def _store_prediction_result(self, result):
+        self._ceph_context._prediction_result = result
+
+    def _parse_prediction_data(self, host_domain_id, disk_domain_id):
+        result = {}
+
+        sql = "SELECT * FROM \"%s\" WHERE (%s) ORDER BY time DESC LIMIT 1"
+        where = "host_domain_id=\'%s\' AND disk_domain_id=\'%s\'" % (host_domain_id, disk_domain_id)
+        try:
+            query_info = self._command.query_info(sql % (self.__class__.measurement, where))
+            status_code = query_info.status_code
+            if status_code == 200:
+                resp = query_info.json()
+                rc = resp.get('results', [])
+                if rc:
+                    series = rc[0].get('series', [])
+                    if series:
+                        values = series[0].get('values', [])
+                        if not values:
+                            return result
+
+                        disk_id_idx = 0
+                        columns = series[0].get('columns', [])
+                        for _idx, _val in enumerate(columns):
+                            if _val == 'disk_domain_id':
+                                disk_id_idx = _idx
+
+                        if not disk_id_idx:
+                            return result
+
+                        for item in values:
+                            # get prediction key and value from server.
+                            for name, value in zip(columns, item):
+                                # process prediction data
+                                result[name] = value
+            else:
+                resp = query_info.json()
+                if resp.get('error'):
+                    self._logger.error(str(resp['error']))
+        except Exception as e:
+            self._logger.error(str(e))
+        return result
+
+    def _effected_resource(self, osd_id):
+        result = {'pool': [], 'rbd': []}
+
+        obj_api = DB_API(self._ceph_context)
+        pools = obj_api.get_osd_pools()
+        effected_pool = set()
+        effected_rbd = set()
+        for pool_data in pools:
+            pool_name = pool_data.get('pool_name')
+            rbd_list = obj_api.get_rbd_list(pool_name=pool_name)
+            for rbd_data in rbd_list:
+                image_name = rbd_data.get('name')
+                rbd_info = obj_api.get_rbd_info(pool_name, image_name)
+                rbd_pgids = rbd_info.get('pgs', [])
+                for _data in rbd_pgids:
+                    acting = _data.get('acting', [])
+                    if osd_id in acting:
+                        effected_pool.add(pool_name)
+                        effected_rbd.add(image_name)
+
+        result['pool'] = map(lambda x: x, effected_pool)
+        result['rbd'] = map(lambda x: x, effected_rbd)
+        return result
+
+    def _fetch_prediction_result(self):

```

Comment:
This method seems to both scrape new smart data form each device and also query the API (with a SQL query??) for that device to fetch an existing health status (not based on the new smart data, but a previously generated prediction).  Why is the latest smart data queried here?  Can't it just look at self.get("devices") to get the cluster's device list?

Reply:
@liewegas Thank your suggestion. This function purpose is retrieving prediction data from AI engine(local/cloud) and use device command to set life expectancy time. I will simplify this function to do this purpose.